[RNMobile] Add i18n scripts for improving the translation pipeline

fluiddot opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This PR introduces scripts for improving the translation pipeline, in the following sections it's briefly described how they work:

### Extract files from the source map

_Script: [`packages/react-native-editor/bin/extract-files-from-sourcemap.js`](https://github.com/WordPress/gutenberg/blob/224123d044f5aa3d2816e0e0aef3e93c676e5151/packages/react-native-editor/bin/extract-files-from-sourcemap.js)_

This script basically identifies the source code files referenced in a source map file and copies them (including the directory structure) to a target path. We won't use it directly as it will be used in the extract used strings script, although it might be useful in case we want to check what files are being incorporated in a JS bundle.

### Generate POT files

_Script: [`packages/react-native-editor/bin/generate-pot-files.sh`](https://github.com/WordPress/gutenberg/blob/224123d044f5aa3d2816e0e0aef3e93c676e5151/packages/react-native-editor/bin/generate-pot-files.sh)_

Similar as we had in GB-mobile, the purpose of this script is to generate the different POT files which contain the strings referenced in source code files. The files generated are the following ones:
-  `gutenberg-blocks.pot`: Contains strings referenced in `block.json` files.
- `gutenberg-source.pot`: Contains strings referenced in all source code files, this is the file we are generating in the current translation pipeline.
- `gutenberg-used-android.pot`: Contains strings referenced in the Android source-map file (previously extracted via the `extract-files-from-sourcemap` script).
- `gutenberg-used-ios.pot`: Contains strings referenced in the iOS source-map file (previously extracted via the `extract-files-from-sourcemap` script).

Similar to the previous script, this won't be used directly and it's part of the process of extracting the used strings described in the next section.

**NOTE:** The script also prints parse errors detected when generating the POT files, it's important to note that strings from files that can't be parsed won't be extracted so this could be a potential cause of missed translations. It's expected that some files can't be parsed due to their structure, however, we should be vigilant on those that we know for sure that contain strings.

### Extract used strings

_Script: [`packages/react-native-editor/bin/extract-used-strings.js`](https://github.com/WordPress/gutenberg/blob/224123d044f5aa3d2816e0e0aef3e93c676e5151/packages/react-native-editor/bin/extract-used-strings.js)_

This is the main script to improve the translation pipeline, it contains a structure to list all the used strings, grouped by domain (in case we use multiple plugins), and provides the following properties for each one:
- `string`: The value of the string.
- `stringPlural`: Plural form value of the string [This parameter is optional].
- `comments`: Comments for translators [the default value is an empty string].
- `reference`: An array containing the paths of the source files that reference the string.
- `platforms`: An array containing the platforms where the string is being used, the values are:
    - `android`: If the string is being used in the Android JS bundle.
    - `ios:` If the string is being used in the iOS JS bundle.
    - `web`: If the string is also being used in the web version of the editor.

### Download translations

_Script: [`packages/react-native-editor/bin/i18n-translations-download.js`](https://github.com/WordPress/gutenberg/blob/224123d044f5aa3d2816e0e0aef3e93c676e5151/packages/react-native-editor/bin/i18n-translations-download.js)_

Additionally, we're providing a unified script for download and optimizing translations. The download logic is very similar to what we previously had but it has been incorporated a new argument for passing a used strings JSON file, which will be used to filter out the unused strings from the translation files.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### 1- Verify extract files from source map script

**Preparation:**
1. Run `npm run native bundle`.

**iOS JS bundle:**
1. Run `mkdir -p local-test/source-files/ios`
1. Run `node packages/react-native-editor/bin/extract-files-from-sourcemap packages/react-native-editor/bundle/ios/App.js.map local-test/source-files/ios`.
1. Navigate to `local-test/source-files/ios` and verify that contains all the source files referenced by the iOS JS bundle. Since there are a lot of files, focus on those files that we know for sure that should be included like the native version of a block.
1. Verify that no `.android.js` files were included ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/picker/index.android.js)).

**Android JS bundle:**
1. Run `mkdir -p local-test/source-files/android`
1. Run `node packages/react-native-editor/bin/extract-files-from-sourcemap packages/react-native-editor/bundle/android/App.text.js.map local-test/source-files/android`.
1. Navigate to `local-test/source-files/android` and verify that contains all the source files referenced by the iOS JS bundle. Since there are a lot of files, focus on those files that we know for sure that should be included like the native version of a block.
1. Verify that no `.ios.js` files were included ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/picker/index.ios.js)).

### 2 - Verify generate POT files script
1. Run `npm run build`.
1. Run `METRO_CONFIG=packages/react-native-editor/metro.config.js packages/react-native-editor/bin/generate-pot-files.sh --path local-test`.
**NOTE:** If you experienced an error related to `react-native-url-polyfill`, run `rm -rf packages/react-native-editor/bundle`.
1. Navigate to `local-test` folder and verify that the following files/directories were created:
    - `bundle/android/App.text.js.map`: Android JS bundle
    - `bundle/ios/App.js.map`: iOS JS bundle
    - `source-files/android`: Should contain all the source files referenced by the Android JS bundle
    - `source-files/ios`: Should contain all the source files referenced by the iOS JS bundle
    - `gutenberg-block.pot`: Should contain all strings referenced by `block.json` files.
    - `gutenberg-source.pot`: Should contain all strings referenced by source files located under `packages`.
    - `gutenberg-used-android.pot`: Should contain all strings referenced by the Android JS bundle.
    - `gutenberg-used-ios.pot`: Should contain all strings referenced by the iOS JS bundle.

**NOTE:** It's expected that some files log parse errors (see attached log), however, we shouldn't get any from files that we know sure that contain strings.

```
Extract used strings from Android source-map:
Success: POT file successfully generated!

Extract used strings from iOS source-map:
Success: POT file successfully generated!

Extract strings from block JSON files:
Plugin file detected.
Success: POT file successfully generated!

Extract strings from non-native JS code:
Plugin file detected.
Success: POT file successfully generated!
Warning: Could not parse file build/components/index.js: Unexpected: 05 (line 48743, column 1847) (17.094s)
Warning: Could not parse file packages/block-editor/src/hooks/duotone.js: Unterminated string (line 103, column 0) (30.089s)
Warning: Could not parse file packages/create-block/index.js: Unexpected # (line 1, column 0) (47.136s)
Warning: Could not parse file packages/i18n/tools/pot-to-php.js: Unexpected # (line 1, column 0) (55.316s)
```

### 3 - Verify extract used strings script
1. Run `METRO_CONFIG=packages/react-native-editor/metro.config.js node packages/react-native-editor/bin/extract-used-strings local-test/used-strings.json`
**NOTE:** If you experienced an error related to `react-native-url-polyfill`, run `rm -rf packages/react-native-editor/bundle`.
1. Navigate to `local-test` folder and verify that file `used-strings.json` contains the used strings.

### 4 - Verify i18n translations download script

**Preparation:**
This test requires the previously created used strings JSON file, which should be located at `local-test/used-strings.json` after running the previous test.

1. Run `node packages/react-native-editor/bin/i18n-translations-download gutenberg local-test/i18n-cache-full`.
1. Navigate to `local-test/i18n-cache-full` folder and verify that translations and `index.js` files are created.
1. Run `node packages/react-native-editor/bin/i18n-translations-download gutenberg local-test/i18n-cache-optimized local-test/used-strings.json`
1. Navigate to `local-test/i18n-cache-optimized` folder and verify that translations and `index.js` files are created.
1. Compare the total size of both folders and verify that optimized translations are smaller.

**NOTE:** The following warnings will be displayed when generating the optimized translations:
```
WARNING: The following strings are missing from translations:
[
  'Pixels (px)',
  'Percentage (%)',
  'Percent (%)',
  'Relative to parent font size (em)',
  'Relative to root font size (rem)',
  'Viewport width (vw)',
  'Viewport height (vh)',
  'Viewport smallest dimension (vmin)',
  'Viewport largest dimension (vmax)',
  'Width of the zero (0) character (ch)',
  'x-height of the font (ex)',
  'Centimeters (cm)',
  'Millimeters (mm)',
  'Inches (in)',
  'Picas (pc)',
  'Points (pt)',
  'Linear',
  'Radial',
  'Angle',
  'Newest to oldest',
  'Oldest to newest',
  'A → Z',
  'Z → A',
  'Order by',
  'Relative to parent font size (em)\x04ems',
  'Relative to root font size (rem)\x04rems'
]
```

```
WARNING: The following native-only strings are already being included in translations:
[ 'Text color' ]
```

Looks like that for the missing ones we need to also download the translations from WordPress ([reference](https://translate.wordpress.org/projects/wp/)), as those strings are translated in that GlotPress project, I'll try to cover those cases in another PR.

Here is an example:
- String: `Radial`
- [Reference in WP GlotPress project](https://translate.wordpress.org/projects/wp/5.8.x/es/default/?filters%5Bterm%5D=Radial&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc)

## Screenshots <!-- if applicable -->
N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

fluiddot committed:
Add extract files from sourcemap script

fluiddot committed:
Add extract used strings script

fluiddot committed:
Add generate pot files script

fluiddot committed:
Add i18n translations download script

fluiddot committed:
Add gettext-parser package to react-native-editor

fluiddot committed:
Add context property to used strings generation

fluiddot commented on packages/react-native-editor/package.json:
```diff
@@ -51,6 +51,7 @@
 		"@wordpress/react-native-bridge": "file:../react-native-bridge",
 		"core-js": "^3.19.1",
 		"fast-average-color": "^4.3.0",
+		"gettext-parser": "^1.3.1",
```
This dependency is required by the `extract-used-strings` script. I also noticed that it's already being used on other packages like `i18n` ([reference](https://github.com/WordPress/gutenberg/blob/a92586c6ba83e735e102867bc69584b6af856643/packages/i18n/package.json#L34)).

fluiddot committed:
Add usage and extra info to scripts

fluiddot committed:
Propagate environment variables in extract used strings

fluiddot committed:
Update Gutenberg source code dir variable

fluiddot committed:
Remove build gutenberg step from generate pot files script

fluiddot committed:
Fix RN bundle cli path

fluiddot commented on packages/react-native-editor/bin/i18n-translations-download.js:
```diff
@@ -0,0 +1,269 @@
+						translationRelativePath
+					);
+
+					// Detect potential missing strings from translations
+					missingTranslations = Object.values( usedStrings ).reduce(
+						( result, { string, platforms } ) => {
+							const stringExists = !! languageResult.response[
+								string
+							];
+							const isMissingString =
+								! stringExists && platforms.includes( 'web' );
+
+							return isMissingString &&
+								! result.includes( string )
+								? [ ...result, string ]
+								: result;
+						},
+						missingTranslations
+					);
```
This will allows us to detect if any used string that should be present in the translations is missing.

fluiddot commented on packages/react-native-editor/bin/i18n-translations-download.js:
```diff
@@ -0,0 +1,269 @@
+						missingTranslations
+					);
+
+					// Detect potential native-only strings that are already included in translations
+					extraTranslations = Object.values( usedStrings ).reduce(
+						( result, { string, platforms } ) => {
+							const stringExists = !! languageResult.response[
+								string
+							];
+							const isNativeTranslationIncluded =
+								! platforms.includes( 'web' ) && stringExists;
+
+							return ! result.includes( string ) &&
+								isNativeTranslationIncluded
+								? [ ...result, string ]
+								: result;
+						},
+						extraTranslations
+					);
```
This part will allow us to detect if any native-only used string is also included in the translations, this is not expected as this type of string is forwarded to the localization strings files.

fluiddot commented on packages/react-native-editor/bin/i18n-translations-download.js:
```diff
@@ -0,0 +1,269 @@
+						extraTranslations
+					);
+
+					// Process translations and filter out unused strings if provided
+					const translationData =
+						Object.keys( usedStrings ).length > 0
+							? Object.keys( languageResult.response )
+									.filter(
+										( string ) => !! usedStrings[ string ]
+									)
+									.reduce(
+										( result, string ) => ( {
+											...result,
+											[ string ]:
+												languageResult.response[
+													string
+												],
+										} ),
+										{}
+									)
+							: languageResult.response;
```
If used strings JSON file is provided, this part will apply it and filter out unused strings.

fluiddot commented on packages/react-native-editor/bin/i18n-translations-download.js:
```diff
@@ -0,0 +1,269 @@
+					// Write translations to file
+					fs.writeFile(
+						translationAbsolutePath,
+						JSON.stringify( translationData, null, 2 ),
```
I decided to prettty-print the translation JSON files to make them more readable, this won't impact the final JS bundle size, as this content is minified when the bundle is generated.

fluiddot commented on packages/react-native-editor/bin/i18n-translations-download.js:
```diff
@@ -0,0 +1,269 @@
+		pluginDir,
+		usedStrings,
+	} ).then( ( { translations, missingTranslations, extraTranslations } ) => {
+		if ( missingTranslations.length > 0 ) {
+			console.log(
+				'WARNING: The following strings are missing from translations:'
+			);
+			console.log( missingTranslations );
+		}
+
+		if ( extraTranslations.length > 0 ) {
+			console.log(
+				'WARNING: The following native-only strings are already being included in translations:'
+			);
+			console.log( extraTranslations );
+		}
```
I was considering aborting the process in case any of these warnings are detected, however, we can't do it as we already have warnings (see PR's description).

fluiddot commented on packages/react-native-editor/bin/extract-files-from-sourcemap.js:
```diff
@@ -0,0 +1,128 @@
+			file.indexOf( 'node_modules' ) === -1 &&
+			file.endsWith( '.ts' )
+		) {
+			const compiledFile = file
+				.replace( 'src', 'build' )
+				.replace( '.ts', '.js' );
```
It's expected that the compiled JS file is located at the same level but under the `build` folder.

fluiddot commented on packages/react-native-editor/bin/extract-files-from-sourcemap.js:
```diff
@@ -0,0 +1,128 @@
+	sourceFiles = sourceFiles.map( ( file ) => {
+		if (
+			// For now, only check files under Gutenberg
+			file.startsWith( './gutenberg' ) !== -1 &&
```
In the future, if other plugins use TypeScript, we should consider expanding this logic.

fluiddot commented on packages/react-native-editor/bin/extract-files-from-sourcemap.js:
```diff
@@ -0,0 +1,128 @@
+
+	// Get source files from sourcemap file
+	const mapFileData = fs.readFileSync( mapFile );
+	const { sources } = JSON.parse( mapFileData );
```
The source code files are extracted from `sources` property, as described in the [sourcemap documentation](https://sourcemaps.info/spec.html).

fluiddot commented on packages/react-native-editor/bin/extract-used-strings.js:
```diff
@@ -0,0 +1,179 @@
+	usedStrings = sourceStrings
+		.filter(
+			( { string, reference } ) =>
+				!! usedStrings[ string ] &&
+				! reference.every( ( ref ) => nativeRefRegex.test( ref ) )
```
For calculating the strings that are also used in the web version of the editor, we check if the string is being used and if none of its references is a native-only file.

fluiddot commented on packages/react-native-editor/bin/extract-used-strings.js:
```diff
@@ -0,0 +1,179 @@
+		.map( ( { name, sourcePath } ) => `${ name } ${ sourcePath }` )
+		.join( ' ' );
+	try {
+		// TODO: Detect potential parse failures
```
Ideally, we should detect from the output of the `generate-pot-files` script potential parse failures. Currently, these failures are only logged when using the debug mode so they're silently skipped.

fluiddot commented on packages/react-native-editor/bin/generate-pot-files.sh:
```diff
@@ -0,0 +1,243 @@
+    $WP_CLI cli update --nightly --yes
+    $WP_CLI --info
+  fi
+}
```
Now the WP-CLI command is only managed by this script, so it will be no longer referenced in the NPM commands.

fluiddot commented on packages/react-native-editor/bin/generate-pot-files.sh:
```diff
@@ -0,0 +1,243 @@
+  echo -e "\nExtract used strings from iOS source-map:"
+  $MAKEPOT_COMMAND $IOS_EXTRACT_SOURCE_FILES_PATH $DEBUG_PARAM $SUBTRACT_PARAM $DOMAIN_PARAM $OUTPUT_POT_USED_IOS_FILE
+
+  echo -e "\nExtract strings from block JSON files:"
+  $MAKEPOT_COMMAND $SOURCE_DIR $DEBUG_PARAM --exclude="$EXCLUDE_FILES" --skip-js --skip-php --ignore-domain $OUTPUT_POT_BLOCKS_FILE
```
Usually, the block JSON files are being parsed if they're contained in the source directory passed, however, I noticed that this is only the case when ignoring the domain. For this reason, I decided to extract it as an extra step and merge it when extracting the strings from non-native JS code.

fluiddot commented on packages/react-native-editor/bin/generate-pot-files.sh:
```diff
@@ -0,0 +1,243 @@
+fi
+
+# Define constants
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
```
This variable is used to get the path of the script for calculating the path of other resources.

fluiddot commented on packages/react-native-editor/bin/generate-pot-files.sh:
```diff
@@ -0,0 +1,243 @@
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+GUTENBERG_SOURCE_CODE_DIR="$SCRIPT_DIR/../../.."
+WP_CLI="php -d memory_limit=4G $SCRIPT_DIR/wp-cli.phar"
+BUNDLE_CLI="$GUTENBERG_SOURCE_CODE_DIR/node_modules/.bin/react-native bundle --config $METRO_CONFIG"
```
I experienced different errors when using the NPM commands for generating the bundle, usually due to different Metro configs and the when is called from GB-mobile. For this reason, I decided to reference the binary from the `node_modules` folder and use an environment variable `METRO_CONFIG` to provide a custom Metro configuration.

fluiddot committed:
Rewrite error message from extract files from sourcemap

fluiddot committed:
Merge branch 'trunk' into rnmobile/add/i18n-scripts

jhnstn commented on packages/react-native-editor/bin/extract-used-strings.js:
```diff
@@ -0,0 +1,179 @@
+const createString = ( previousString, string, platform ) =>
+	previousString
+		? {
+				...previousString,
```
Is the indentation off here?

fluiddot replied:
I reformatted the file and looks like the indentation is correct, is producing a different result on your side?

jhnstn replied:
OK.  Seems like there is too many tabs in there but it's probably just the ternary formatting that pushes out `...previousString` , Thanks for checking!

fluiddot committed:
Rename local variables of generate pot files script

fluiddot committed:
Add more options to make i18 update script safer

fluiddot committed:
Merge branch 'trunk' into rnmobile/add/i18n-scripts

jhnstn reviewed (changes requested):
I'm getting an unbound variable error when trying to perform the `extract-used-strings` step 

```/Users/jason/Projects/wordpress-mobile/gutenberg-mobile/gutenberg/packages/react-native-editor/bin/generate-pot-files.sh: line 243: PLUGINS_TO_EXTRACT_FROM_GUTENGERG[@]: unbound variable```

fluiddot committed:
Fix unbound variable errors

fluiddot commented:
> I'm getting an unbound variable error when trying to perform the `extract-used-strings` step
> 
> `/Users/jason/Projects/wordpress-mobile/gutenberg-mobile/gutenberg/packages/react-native-editor/bin/generate-pot-files.sh: line 243: PLUGINS_TO_EXTRACT_FROM_GUTENGERG[@]: unbound variable`

Good catch, looks like it's caused when not adding extra plugins to the command. I'm afraid I overlooked this error because I was testing the `generate-pot-files.sh` script via the new i18n update script from GB-mobile ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366)) which doesn't produce this issue.

I fixed the issue in https://github.com/WordPress/gutenberg/pull/37482/commits/fd365cc1d51b87e3a9e6fb0aa7e2dec6523ebfb9.

fluiddot committed:
Detect parse errors when generating pot files

Additionally, a "make_pot" function has been created to simplify the command.

fluiddot committed:
Add quotes to Typescript file build log message

fluiddot committed:
Use root path of Gutenberg for generating pot files

Previously we were pointing to "packages" folder, which led to overlook some files that contain strings.

fluiddot committed:
Handle strings without reference case

fluiddot committed:
Update generate pot files log message

fluiddot commented:
@jhnstn I've just pushed a couple of changes for addressing the following items:
1. Detect potential parse errors when generating POT files (https://github.com/WordPress/gutenberg/pull/37482/commits/680f156efa71e24b6b980e210d44ef5c79d65743): This will help us on identifying potentially missed strings if we spot that files are not properly being parsed.
2. Use Gutenberg's root path for extracting strings (https://github.com/WordPress/gutenberg/pull/37482/commits/74d725acf976dd77f97f34eccd35c513091af7d1): This will fix the warning for the string `Featured` which is referenced by a PHP file in a different folder than `packages` (the path we were using previously).

fluiddot commented on packages/react-native-editor/bin/generate-pot-files.sh:
```diff
@@ -0,0 +1,254 @@
+  local full_command="$makepot_command $source --debug $arguments"
+  if [[ -z ${DEBUG:-} ]]; then
+    # When DEBUG flag is not enabled, we only print the parse errors.
+    $full_command 2> >(grep 'Could not parse file' | sed 's/Debug[[:space:]][(]make-pot[)]/Warning/g' >&2)
```
The parse errors are logged by the i18n command to `stderr` as:
`Debug (parse-pot): Could not parse file <FILE>`

This line substitutes the content of `stderr` by only including the parse errors entries and replacing the title `Debug (parse-pot)` with `Warning`.

jhnstn reviewed (approved):
I ran through the test steps without any issues. I also spot checked the results of test commands and everything looks correct as far as I can tell.

Nice work!
